File: StrukturDataQueue/Queue.py
import tkinter as tk
from tkinter import ttk, messagebox
import pyttsx3
import threading
import time

# DATA
queue = []
nomor = 1


# TTS
def speak(text):
    def run():
        engine = pyttsx3.init()
        engine.setProperty('rate', 150)
        engine.say(text)
        engine.runAndWait()
    threading.Thread(target=run).start()


# FUNCTION
def ambil_antrian():
    global nomor
    nama = entry_nama.get()

    if nama == "":
        messagebox.showwarning("Warning", "Nama harus diisi!")
        return

    queue.append((nomor, nama))
    listbox.insert("", "end", values=(nomor, nama))


    nomor += 1
    entry_nama.delete(0, tk.END)


def panggil_antrian():
    if len(queue) == 0:
        messagebox.showinfo("Info", "Antrian kosong")
        return

    no, nama = queue.pop(0)
    listbox.delete(listbox.get_children()[0])

    label_nomor.config(text=str(no))
    label_nama.config(text=nama)

    speak(f"Nomor antrian {no}, atas nama {nama}, silakan menuju loket.")


def tampilkan_data():
    if len(queue) == 0:
        messagebox.showinfo("Info", "Antrian kosong")
    else:
        data = "\n".join([f"{n} - {nm}" for n, nm in queue])
        messagebox.showinfo("Data Antrian", data)


# GUI SETUP
root = tk.Tk()
root.title("Antrian Bank")
root.geometry("600x550")
root.configure(bg="#eef5ff")

style = ttk.Style()
style.theme_use("default")

# STYLE BUTTON
style.configure("TButton",
                font=("Segoe UI", 10, "bold"),
                padding=10)

# HEADER
header = tk.Label(root, text="SISTEM ANTRIAN BANK",
                  font=("Segoe UI", 18, "bold"),
                  bg="#eef5ff", fg="#0d47a1")
header.pack(pady=10)

# DISPLAY BESAR
frame_display = tk.Frame(root, bg="#eef5ff")
frame_display.pack(pady=10)

label_nomor = tk.Label(frame_display, text="-",
                       font=("Segoe UI", 50, "bold"),
                       fg="#1565c0", bg="#eef5ff")
label_nomor.pack()

label_nama = tk.Label(frame_display, text="-",
                      font=("Segoe UI", 14),
                      bg="#eef5ff")
label_nama.pack()

# INPUT
frame_input = tk.Frame(root, bg="#eef5ff")
frame_input.pack(pady=10)

tk.Label(frame_input, text="Nama:", bg="#eef5ff").grid(row=0, column=0, padx=5)
entry_nama = ttk.Entry(frame_input, width=25)
entry_nama.grid(row=0, column=1, padx=5)

# BUTTON
frame_btn = tk.Frame(root, bg="#eef5ff")
frame_btn.pack(pady=10)

ttk.Button(frame_btn, text="Ambil Antrian", command=ambil_antrian).grid(row=0, column=0, padx=10)
ttk.Button(frame_btn, text="Panggil Antrian", command=panggil_antrian).grid(row=0, column=1, padx=10)
ttk.Button(frame_btn, text="Tampilkan Data", command=tampilkan_data).grid(row=0, column=2, padx=10)

# TABLE
columns = ("Nomor", "Nama")
listbox = ttk.Treeview(root, columns=columns, show="headings", height=10)

listbox.heading("Nomor", text="Nomor")
listbox.heading("Nama", text="Nama")

listbox.column("Nomor", width=80, anchor="center")
listbox.column("Nama", width=200)

listbox.pack(pady=10)

# RUN
root.mainloop()

File: StrukturDataQueue/Queue.java
import com.formdev.flatlaf.FlatDarkLaf;

import javax.swing.*;
import java.awt.*;
import java.util.LinkedList;

public class Queue extends JFrame {

    static class Node {
        int nomor;
        String nama;

        Node(int nomor, String nama) {
            this.nomor = nomor;
            this.nama = nama;
        }
    }

    LinkedList<Node> queue = new LinkedList<>();
    int nomor = 1;

    JLabel lblNomor, lblNama;
    JTextArea area;
    JTextField inputNama;

    public Queue() {
        setTitle("ANTRIAN BANK");
        setSize(900, 650);
        setLocationRelativeTo(null);
        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setResizable(false);

        // MAIN PANEL
        JPanel mainPanel = new JPanel(new BorderLayout(20, 20));
        mainPanel.setBorder(BorderFactory.createEmptyBorder(20, 20, 20, 20));
        
        // HEADER
        JLabel header = new JLabel("ANTRIAN BANK", SwingConstants.CENTER);
        header.setFont(new Font("Segoe UI", Font.BOLD, 40));
        mainPanel.add(header, BorderLayout.NORTH);

        // CENTER - DISPLAY PANEL
        JPanel displayPanel = new JPanel(new BorderLayout(20, 20));
        displayPanel.setBorder(BorderFactory.createTitledBorder("Sedang Dipanggil"));
        
        lblNomor = new JLabel("-");
        lblNomor.setFont(new Font("Segoe UI", Font.BOLD, 140));
        lblNomor.setHorizontalAlignment(SwingConstants.CENTER);
        
        lblNama = new JLabel("Menunggu...");
        lblNama.setFont(new Font("Segoe UI", Font.PLAIN, 28));
        lblNama.setHorizontalAlignment(SwingConstants.CENTER);
        
        displayPanel.add(lblNomor, BorderLayout.CENTER);
        displayPanel.add(lblNama, BorderLayout.SOUTH);
        
        mainPanel.add(displayPanel, BorderLayout.CENTER);

        // SOUTH - CONTROL PANEL
        JPanel controlPanel = new JPanel(new BorderLayout(15, 15));
        
        // Input panel
        JPanel inputPanel = new JPanel(new BorderLayout(10, 0));
        inputNama = new JTextField(25);
        inputNama.setFont(new Font("Segoe UI", Font.PLAIN, 14));
        inputPanel.add(new JLabel("Nama:"), BorderLayout.WEST);
        inputPanel.add(inputNama, BorderLayout.CENTER);
        controlPanel.add(inputPanel, BorderLayout.NORTH);

        // Button panel
        JPanel buttonPanel = new JPanel(new GridLayout(1, 3, 15, 0));
        JButton btnAmbil = new JButton("Ambil Antrian");
        JButton btnTampil = new JButton("Tampilkan");
        JButton btnPanggil = new JButton("Panggil");
        
        btnAmbil.setFont(new Font("Segoe UI", Font.BOLD, 14));
        btnTampil.setFont(new Font("Segoe UI", Font.BOLD, 14));
        btnPanggil.setFont(new Font("Segoe UI", Font.BOLD, 14));
        
        btnAmbil.setFocusPainted(false);
        btnTampil.setFocusPainted(false);
        btnPanggil.setFocusPainted(false);
        
        buttonPanel.add(btnAmbil);
        buttonPanel.add(btnTampil);
        buttonPanel.add(btnPanggil);
        controlPanel.add(buttonPanel, BorderLayout.CENTER);

        // Area
        area = new JTextArea(8, 40);
        area.setFont(new Font("Monospaced", Font.PLAIN, 13));
        area.setEditable(false);
        area.setLineWrap(true);
        
        JScrollPane scrollArea = new JScrollPane(area);
        scrollArea.setBorder(BorderFactory.createTitledBorder("Daftar Antrian"));
        controlPanel.add(scrollArea, BorderLayout.SOUTH);
        
        mainPanel.add(controlPanel, BorderLayout.SOUTH);

        add(mainPanel);

        // EVENT LISTENERS
        btnAmbil.addActionListener(e -> ambil());
        btnTampil.addActionListener(e -> tampil());
        btnPanggil.addActionListener(e -> panggil());
    }

    private void ambil() {
        String nama = inputNama.getText().trim();
        if (nama.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Nama tidak boleh kosong!", "Warning", JOptionPane.WARNING_MESSAGE);
            return;
        }
        queue.add(new Node(nomor, nama));
        nomor++;
        inputNama.setText("");
        tampil();
    }

    private void tampil() {
        StringBuilder sb = new StringBuilder();
        if (queue.isEmpty()) {
            sb.append("[Antrian Kosong]");
        } else {
            int index = 1;
            for (Node n : queue) {
                sb.append(String.format("%d. No %d - %s%n", index++, n.nomor, n.nama));
            }
        }
        area.setText(sb.toString());
    }

    private void panggil() {
        if (queue.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Antrian kosong!", "Info", JOptionPane.INFORMATION_MESSAGE);
            return;
        }
        
        Node n = queue.poll();
        lblNomor.setText(String.valueOf(n.nomor));
        lblNama.setText(n.nama);
        
        // Jalankan text-to-speech di thread terpisah
        new Thread(() -> {
            speakAI("Nomor " + n.nomor + ", atas nama " + n.nama + ", silakan ke loket");
        }).start();
        
        System.out.println(">>> Memanggil: No " + n.nomor + " - " + n.nama);
        
        tampil();
    }

    private void speakAI(String text) {
        new Thread(() -> {
            try {
                // Coba pakai Windows Native Speech Synthesizer
                // Dengan voice Indonesia jika tersedia
                String[] cmd = {
                    "powershell.exe",
                    "-Command",
                    "Add-Type -AssemblyName System.Speech; " +
                    "$speak = New-Object System.Speech.Synthesis.SpeechSynthesizer; " +
                    "$speak.Rate = -1; " +  // Speed setting
                    "$speak.Volume = 100; " +
                    "$speak.Speak('" + text.replace("'", "''") + "')"
                };
                
                Process p = Runtime.getRuntime().exec(cmd);
                p.waitFor();
                
                System.out.println("TTS executed: " + text);
                
            } catch (Exception e) {
                System.err.println("Error speaking: " + e.getMessage());
                e.printStackTrace();
            }
        }).start();
    }

    public static void main(String[] args) {
        try {
            FlatDarkLaf.install();
        } catch (Exception e) {
            System.err.println("FlatLaf setup failed: " + e.getMessage());
        }

        SwingUtilities.invokeLater(() -> {
            new Queue().setVisible(true);
        });
    }
}
